certdepot/mongo_depot.go

```diff
@@ -0,0 +1,234 @@
+package certdepot
+
+import (
+	"time"
+
+	"github.com/mongodb/anser/bsonutil"
+	"github.com/mongodb/grip"
+	"github.com/mongodb/grip/message"
+	"github.com/pkg/errors"
+	"github.com/square/certstrap/depot"
+	mgo "gopkg.in/mgo.v2"
+	"gopkg.in/mgo.v2/bson"
+)
+
+type User struct {
+	ID            string    `bson:"_id"`
+	Cert          string    `bson:"cert"`
+	PrivateKey    string    `bson:"private_key"`
+	CertReq       string    `bson:"cert_req"`
+	CertRevocList string    `bson:"cert_revoc_list"`
+	TTL           time.Time `bson:"ttl"`
+}
+
+var (
+	userIDKey            = bsonutil.MustHaveTag(User{}, "ID")
+	userCertKey          = bsonutil.MustHaveTag(User{}, "Cert")
+	userPrivateKeyKey    = bsonutil.MustHaveTag(User{}, "PrivateKey")
+	userCertReqKey       = bsonutil.MustHaveTag(User{}, "CertReq")
+	userCertRevocListKey = bsonutil.MustHaveTag(User{}, "CertRevocList")
+	userTTLKey           = bsonutil.MustHaveTag(User{}, "TTL")
+)
+
+type mongoCertDepot struct {
+	session        *mgo.Session
+	databaseName   string
+	collectionName string
+	expireAfter    time.Duration
+}
+
+type MgoCertDepotOptions struct {
+	MongoDBURI           string
+	MongoDBDialTimeout   time.Duration
+	MongoDBSocketTimeout time.Duration
+	DatabaseName         string
+	CollectionName       string
+	ExpireAfter          time.Duration
+}
+
+// Create a new cert depot in the specified MongoDB.
+func NewMgoCertDepot(opts MgoCertDepotOptions) (depot.Depot, error) {
+	return newMgoCertDeposit(nil, opts)
+}
+
+// Create a new cert depot in the specified MongoDB, using an existing session.
+func NewMgoCertDepotWithSession(s *mgo.Session, opts MgoCertDepotOptions) (depot.Depot, error) {
+	return newMgoCertDeposit(s, opts)
+}
+
+func validate(opts MgoCertDepotOptions) MgoCertDepotOptions {
+	if opts.MongoDBURI == "" {
+		opts.MongoDBURI = "mongodb://localhost:27017"
+	}
+	if opts.MongoDBDialTimeout <= 0 {
+		opts.MongoDBDialTimeout = 2 * time.Second
+	}
+	if opts.MongoDBSocketTimeout <= 0 {
+		opts.MongoDBSocketTimeout = time.Minute
+	}
+	if opts.DatabaseName == "" {
+		opts.DatabaseName = "certDepot"
+	}
+	if opts.CollectionName == "" {
+		opts.CollectionName = "certs"
+	}
+	if opts.ExpireAfter <= 0 {
+		opts.ExpireAfter = 30 * 24 * time.Hour
+	}
+	return opts
+}
+
+func newMgoCertDeposit(s *mgo.Session, opts MgoCertDepotOptions) (depot.Depot, error) {
+	opts = validate(opts)
+
+	if s == nil {
+		var err error
+		s, err = mgo.DialWithTimeout(opts.MongoDBURI, opts.MongoDBDialTimeout)
+		if err != nil {
+			return nil, errors.Wrapf(err, "could not connect to db %s", opts.MongoDBURI)
+		}
+		s.SetSocketTimeout(opts.MongoDBSocketTimeout)
+	}
```

Comment:
I'd move the s == nil block to line 51...

Reply:
Why? It goes nicely here, it makes this callable from both `NewMgoCertDepot` and `NewMgoCertDepotWithSession` and in the event a nil session is passed in to `NewMgoCertDepotWithSession` it will connect to the mongo uri in the options. Unless we shouldn't check for a nil session?